pylearn2/datasets/preprocessing.py

```diff
@@ -1421,15 +1427,16 @@ def fit(self, X):
         t1 = time.time()
         eigs, eigv = linalg.eigh(covariance)
         t2 = time.time()
+
         log.info("eigh() took {0} seconds".format(t2 - t1))
         assert not contains_nan(eigs)
         assert not contains_nan(eigv)
         assert eigs.min() > 0
         if self.n_components:
-            eigs = eigs[:self.n_components]
-            eigv = eigv[:, :self.n_components]
+            eigs = eigs[-self.n_components:]
+            eigv = eigv[:, -self.n_components:]
 
-        if self.n_drop_components:
+        if self.n_drop_components and not self.n_components:
             eigs = eigs[self.n_drop_components:]
             eigv = eigv[:, self.n_drop_components:]
```

Comment:
Might be a good idea to raise an error if both `n_components` and `n_drop_components` are not `None`.